src/python/pants/util/objects.py

```diff
@@ -60,6 +60,11 @@ def make_type_error(cls, msg, *args, **kwargs):
       return TypeCheckError(cls.__name__, msg, *args, **kwargs)
 
     def __new__(cls, *args, **kwargs):
+      # TODO: Ideally we could execute this exactly once per `cls` but it should be a
```

Comment:
Can't we just do this in the `datatype` constructor, changing `return type(superclass_name, (DataType,), {})` to:
```
t = type(superclass_name, (DataType,), {})
if not hasattr(t.__eq__, '_eq_override_canary'):
  raise ...
return t
```
?

Reply:
That would check super classes, but not subclasses, I think?

Reply:
Aah, we subclass these things? Got it :)